docs/adr/0004-only-threadsafe-interfaces.md

```diff
@@ -0,0 +1,118 @@
+# Remove support for non-threadsafe interfaces
+
+* Status: proposed
+* Deciders: mhammond, proposed: rfkelly, jhugman, dmose, travis, janerik
+* Date: 2021-03-31
+
+Technical Story: [Issue 419](https://github.com/mozilla/uniffi-rs/issues/419)
+
+## Context and Problem Statement
+
+[ADR-0003](0003-threadsafe-interfaces.md) introduced support for "thread-safe
+interfaces" - possibly leading to the impression that there is such a thing as
+non-threadsafe interfaces.
+
+However, the entire concept of non-threadsafe interfaces is a mismomer -
+everything wrapped by uniffi is thread-safe - the only question is who manages
+this thread-safety. Interfaces which are not marked as thread-safe cause uniffi
+to wrap the interface in a mutex which is hidden in the generated code and
+therefore not obvious to the casual reader.
+
+## Decision Drivers
+
+* Non-threadsafe structs are considered a "foot-gun", leading to accidentally
+  introduce issues like [this Fenix bug](https://github.com/mozilla-mobile/fenix/issues/17086)
+  (with more details available in [this JIRA ticket](https://jira.mozilla.com/browse/SDK-157).)
+
+* Supporting such structs will hinder uniffi growing in directions that we've
+  found are desired in practice, such as allowing structs to use [alternative
+  method receivers](https://github.com/mozilla/uniffi-rs/issues/417) or to
+  [pass interface references over the FFI](https://github.com/mozilla/uniffi-rs/issues/419).
+
+## Considered Options
+
+* [Option 1] Continue supporting non-threadsafe interfaces while also working
+  on the enhancements listed above, but exclude non-threadsafe interfaces from
+  such enhancements.
+
+* [Option 2] Immediately deprecate, then remove entirely, support for
+  non-threadsafe interfaces.
+
+## Decision Outcome
+
+Chosen option:
+
+* **[Option 2] Immediately deprecate, then remove entirely, support for
+  non-threadsafe interfaces.**
+
+This decision was taken because our real world experience tells us that
+non-threadsafe interfaces are only useful in toy or example applications (eg,
+the nimbus and autofill projects didn't get very far before needing these
+capabilities), so the extra ongoing work in supporting these interfaces can not
+be justified.
+
+### Positive Consequences
+
+* The locking in all uniffi supported applications will be clear.
```

Comment:
I endorse "clear", but I also like to describe it as "easily discoverable" - that is, you can look at your code and its type signatures and correctly understand what the locking/blocking behaviour will be without have to understand anything about the internals of uniffi.

Reply:
> Which is to say that I'd vote for also dropping the [Threadsafe] UDL attribute as well (without changing any
> other semantics).

I agree, and in fact I had implicitly assumed that this is what would happen - the `[Threadsafe]` attribute would disappear and its requirement that the underlying Rust struct be `Sync + Send` would apply by default to all interfaces exposed via UniFFI. It certainly seems worth making that more explicit in the doc here.

Reply:
> I agree, and in fact I had implicitly assumed that this is what would happen - the `[Threadsafe]` attribute would disappear and its requirement that the underlying Rust struct be `Sync + Send` would apply by default to all interfaces exposed via UniFFI. It certainly seems worth making that more explicit in the doc here.

I'm not sure where that would naturally live TBH - I don't think here is the right place. I actually thought "remove entirely, support for non-threadsafe interfaces" is fairly clear in that regard.

Reply:
> I'm not convinced that this will, by itself, cause most casual SDK users to reason correctly about what's likely to block and what's not.

Yes, fair point - I'll reword this to say something like "less hidden" which is actually the point being made.

Reply:
We could say a few more explicit words about it under the "implications" section maybe, describing the intended final state along the lines of the above "`[Threadsafe]` disappears, all interfaces `Sync + Send`" as above.

Reply:
heh - actually, `"[Threadsafe] disappears` probably isn't correct as that would break existing code in ways we don't want to. It will be more like "Threadsafe becomes the default, and eventually the attribute becomes deprecated" or something. I'll try and come up with something suitable in the "implications" section.